crates/scroll/consensus/src/validation.rs

```diff
@@ -189,3 +314,66 @@ fn validate_against_parent_timestamp<H: BlockHeader>(
     }
     Ok(())
 }
+
+/// Validates the gas limit of the block against the parent.
+#[inline]
+fn validate_against_parent_gas_limit<H: BlockHeader>(
+    header: &H,
+    parent: &H,
+) -> Result<(), ConsensusError> {
+    let diff = header.gas_limit().abs_diff(parent.gas_limit());
+    let limit = parent.gas_limit().saturating_div(GAS_LIMIT_BOUND_DIVISOR);
+    if diff > limit {
+        return if header.gas_limit() > parent.gas_limit() {
+            Err(ConsensusError::GasLimitInvalidIncrease {
+                parent_gas_limit: parent.gas_limit(),
+                child_gas_limit: parent.gas_limit(),
+            })
+        } else {
+            Err(ConsensusError::GasLimitInvalidDecrease {
+                parent_gas_limit: parent.gas_limit(),
+                child_gas_limit: parent.gas_limit(),
+            })
+        }
+    }
+    // Check that the gas limit is above the minimum allowed gas limit.
+    if header.gas_limit() < MINIMUM_GAS_LIMIT {
+        return Err(ConsensusError::GasLimitInvalidMinimum { child_gas_limit: header.gas_limit() })
+    }
+    Ok(())
+}
+
+/// Validate the L1 messages by checking they are only present that the start of the block and only
+/// have increasing queue index.
+#[inline]
+fn validate_l1_messages<Tx: SignedTransaction + ScrollTransaction>(
+    txs: &[Tx],
+) -> Result<(), ScrollConsensusError> {
+    // Check if the block contains L1 messages.
+    if !txs.iter().any(ScrollTransaction::is_l1_message) {
+        return Ok(())
+    }
```

Comment:
Doesn't this check ultimately imply that if there is an L1 message present in the block, then we end up iterating transactions twice? Is there any value in this?